### PR TITLE
chore!: merge `fuels-types` and `fuels-core`

### PR DESCRIPTION
**Pull request by hal3e**
closes: https://github.com/FuelLabs/fuels-rs/issues/945

`fuels-types` is merged into `fuels-core`. This is the new structure:

```shell
fuels-core
├── Cargo.toml
└── src
    ├── codec
    │   ├── abi_decoder.rs
    │   ├── abi_encoder.rs
    │   └── function_selector.rs
    ├── codec.rs
    ├── lib.rs
    ├── traits
    │   ├── parameterize.rs
    │   └── tokenizable.rs
    ├── traits.rs
    ├── types
    │   ├── bech32.rs
    │   ├── core
    │   │   ├── bits.rs
    │   │   ├── bytes.rs
    │   │   ├── native.rs
    │   │   ├── raw_slice.rs
    │   │   └── sized_ascii_string.rs
    │   ├── core.rs
    │   ├── enum_variants.rs
    │   ├── errors.rs
    │   ├── param_types.rs
    │   ├── transaction_builders.rs
    │   ├── unresolved_bytes.rs
    │   ├── wrappers
    │   │   ├── block.rs
    │   │   ├── chain_info.rs
    │   │   ├── coin.rs
    │   │   ├── input.rs
    │   │   ├── message_proof.rs
    │   │   ├── message.rs
    │   │   ├── node_info.rs
    │   │   ├── resource.rs
    │   │   ├── transaction_response.rs
    │   │   └── transaction.rs
    │   └── wrappers.rs
    ├── types.rs
    ├── utils
    │   ├── constants.rs
    │   └── offsets.rs
    └── utils.rs
```
- modules `utils` and `types::core` and `types::wrappers` are not exposed. They are used for code organization.
- `traits` are moved to `fuels-core` - _**breaking!**_
- `codec` is now used to get to `ABIEncoder`, `ABIDecoder`, `try_from_bytes` and `resolve_fn_selector` - _**breaking!**_
- exposed `fuels_core::types` only through `fuels::types`

### Checklist
- [x] I have linked to any relevant issues.
- [x] I have updated the documentation.
- [x] I have added necessary labels.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

File: packages/fuels-core/src/utils.rs
Comment (segfault-magnet) [inline packages/fuels-core/src/utils.rs:25]: This is also intimately related to stuff over in the new `codec` mod, maybe we should move it there and leave this file empty?

File: packages/fuels-core/src/codec.rs
Comment (hal3e) [reply packages/fuels-core/src/codec.rs:3]: Currently we have `fuels::core::abi_encoder::ABIEncoder`.
Do you mean `fuels::core::codec::ABIEncoder` or `fuels::codec::ABIEncoder` ?

**Comment by iqdecay** [comment]
Haven't reviewed yet, but I suggest we hold off on this PR until #950 is merged because it's more pressing than this tech debt. I should push the corrections to the PR today.

File: packages/fuels-core/src/codec.rs
Comment (segfault-magnet) [reply packages/fuels-core/src/codec.rs]: Sorry, I meant `fuels::core::codec::ABIEncoder`

**Review by segfault-magnet — APPROVED** [review]

**Review by segfault-magnet — COMMENTED** [review]
Reviewed, seems good. 

Waiting for the fuels types reexports to remove redundancy in the import path.

IIRC we agreed that was to be done here. If not ping me and I'll approve.

**Review by segfault-magnet — APPROVED** [review]
Looks good.

I'd give up on the idea of shortening the reexport path since even the std does it like we do (e.g. `std::result::Result`)

**Review by Br1ght0ne — APPROVED** [review]

**Review by iqdecay — APPROVED** [review]
nice reorg, feels much cleaner now :broom:

**Review by digorithm — APPROVED** [review]
Wonderful — ship it!

**Pull request merged**
